csrc/flash_attn_ck/mha_fwd.cpp

```diff
@@ -264,11 +263,24 @@ mha_fwd(at::Tensor &q,                            // batch_size x seqlen_q x num
         p = torch::empty({batch_size, num_heads, seqlen_q, seqlen_k}, opts.dtype(torch::kUInt8));
     }
 
-    uint64_t drop_seed = 1;
-    uint64_t drop_offset = 0;
+    uint64_t drop_seed = 1, drop_offset = 0;
+    int64_t counter_offset = batch_size * num_heads * 64;
     auto options = torch::TensorOptions().dtype(torch::kFloat32).device(torch::kCUDA);
     auto rng_state = torch::empty({2}, options.dtype(torch::kInt64));
-    // TODO - assign seed & offset to rng_state
+
+    if (p_dropout > 0.0)  {
+        auto gen = at::get_generator_or_default<at::CUDAGeneratorImpl>(
+            gen_, at::cuda::detail::getDefaultCUDAGenerator());
+        // See Note [Acquire lock when using random generators]
+        std::lock_guard<std::mutex> lock(gen->mutex_);
+        auto philox_args = gen->philox_cuda_state(counter_offset);
+        auto seed_offset = unpack(philox_args);
+        drop_seed = std::get<0>(seed_offset);
+        drop_offset = std::get<1>(seed_offset);
```

Comment:
you can simply use `std::tie()` to get all the values of `std::tuple<>` :
```c++
std::tie(drop_seed, drop_offset) = unpack(philox_args)
```

Reply:
e8007c772b6d6ad4c62f57663363f35aae0d1063